staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler_test.go

```diff
@@ -638,6 +648,217 @@ func testHandlerConversion(t *testing.T, enableWatchCache bool) {
+		expectedErr           error
+	}{
+		{
+			name:           "strict-decoding",
+			body:           multiVersion,
+			strictDecoding: true,
+			crd:            multiVersionFixture.DeepCopy(),
+			expectedObj:    decoded,
+			expectedErr:    errors.New(`strict decoding error: duplicate field "num", unknown field "unknown"`),
+		},
+		{
+			name:           "non-strict-decoding",
+			body:           multiVersion,
+			strictDecoding: false,
+			crd:            multiVersionFixture.DeepCopy(),
+			expectedObj:    decoded,
+			expectedErr:    nil,
+		},
+		{
+			name:                  "strict-decoding-preserve-unknown",
+			body:                  multiVersion,
+			strictDecoding:        true,
+			preserveUnknownFields: true,
+			crd:                   multiVersionFixture.DeepCopy(),
+			expectedObj:           decodedPreserveUnknown,
+			expectedErr:           errors.New(`strict decoding error: duplicate field "num"`),
+		},
+		{
+			name:                  "non-strict-decoding-preserve-unknown",
+			body:                  multiVersion,
+			strictDecoding:        false,
+			preserveUnknownFields: true,
+			crd:                   multiVersionFixture.DeepCopy(),
+			expectedObj:           decodedPreserveUnknown,
+			expectedErr:           nil,
+		},
+		{
+			name:           "strict-decoding-yaml",
+			body:           multiVersionYAML,
+			yaml:           true,
+			strictDecoding: true,
+			crd:            multiVersionFixture.DeepCopy(),
+			expectedObj:    decoded,
+			expectedErr: errors.New(`strict decoding error: yaml: unmarshal errors:
+  line 7: key "num" already set in map, unknown field "unknown"`),
+		},
+		{
+			name:           "non-strict-decoding-yaml",
+			body:           multiVersionYAML,
+			yaml:           true,
+			strictDecoding: false,
+			crd:            multiVersionFixture.DeepCopy(),
+			expectedObj:    decoded,
+			expectedErr:    nil,
+		},
+		{
+			name:                  "strict-decoding-preserve-unknown-yaml",
+			body:                  multiVersionYAML,
+			yaml:                  true,
+			strictDecoding:        true,
+			preserveUnknownFields: true,
+			crd:                   multiVersionFixture.DeepCopy(),
+			expectedObj:           decodedPreserveUnknown,
+			expectedErr: errors.New(`strict decoding error: yaml: unmarshal errors:
+  line 7: key "num" already set in map`),
+		},
+		{
+			name:                  "non-strict-decoding-preserve-unknown-yaml",
+			body:                  multiVersionYAML,
+			yaml:                  true,
+			strictDecoding:        false,
+			preserveUnknownFields: true,
+			crd:                   multiVersionFixture.DeepCopy(),
+			expectedObj:           decodedPreserveUnknown,
+			expectedErr:           nil,
+		},
+	}
+	for _, tc := range testcases {
+		t.Run(tc.name, func(t *testing.T) {
+			v := "v1beta1"
+
+			var err error
+			structuralSchemas := map[string]*structuralschema.Structural{}
+			structuralSchemas[v], err = generateStructuralSchema(tc.crd, v)
+			if err != nil {
+				t.Fatal(err)
+			}
```

Comment:
If I'm reading correctly and `tc.crd` is always the same, then you could move that out?